### DISCUSSION TITLE
block spammers by domain/IP/email address

### DISCUSSION THREAD

**Discussion by kamarya** [emersion/maddy#356] [Q&A] [answered]
How can I block spammers by domain, IP or email address ? is there any config file or part of the code to modify/patch ?

**Comment by foxcpp** [comment, marked as answer]
https://github.com/foxcpp/maddy/blob/c8659ad18aac206b6b9345926c20d6f49a4daab0/maddy.conf#L93

It is possible to add
```
source DOMAIN { reject }
```
or 
```
source EMAIL { reject }
```
here to reject messages with corresponding sender domain/email.

IP-based blocking is yet to implemented: https://github.com/foxcpp/maddy/issues/297
The way to go for this currently is to use system firewall.

**Reply by kamarya** [reply]
The spammer has an email like `spammer@part1.part2.com` and I added `source part2.com { reject }`. It does not work. What should I do to have a wildcard cause `part1` changes.

**Reply by foxcpp** [reply]
Regular `source` directive matches against the full domain.
You can use `source_in` with `table.regexp` for advanced matching (including wildcards):
`source_in regexp ".+@.+\.part2\.com"`

**Reply by Ramblurr** [reply]
@foxcpp Is this still supposed to work in 0.6.3? I'm trying the following, but the `default_source` is always matching instead

```
smtp tcp://0.0.0.0:25 {
    debug true
    io_debug true
    source_in regexp ".*@.*\.example.com" {
        deliver_to &remote_queue
    }
    default_source {
        reject 521 5.0.0 "Domain not local"
    }
}
```

The idea is to allow email where the sender has any subdomain of `example.com`. But I'm getting `sender root@some.test.example.com matched by default rule	`

**Comment by jtgans** [comment]
So I have a hacky way of doing this in my maddy.conf:

```
smtp tcp://0.0.0.0:25 {
    modify {
        replace_sender regexp "(.+)@(.+)\.click" "spammer@blacklisted.spammer"
    }
    source blacklisted.spammer {
        reject 541 5.4.0 "H8U"
    }
}
```

This is a kind of horrible hack, but you can add as many `replace_sender regexp ...` lines to the `modify` section to block an arbitrary number of domains. In my case, I'm blocking off the new-generation TLDs I never do business with such as click, today, life, etc. In your case, it would just be a matter of using the appropriate regexp.

@foxcpp `source_in regexp ...` seems to be rather broken -- I had originally tried to use that instead, but it's not clear how to use it properly without accidentally rejecting all incoming mail. A solid example of how to use it for a blacklist would go a long way.
